src/transformers/trainer.py

```diff
@@ -1170,6 +1186,122 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
+                if not isinstance(module, nn.Linear):
+                    continue
+
+                if not check_target_module_exists(args.optim_target_modules, module_name) and not all_linear:
+                    continue
+
+                galore_params.append(module.weight)
+                galore_params_names.append(module_name + ".weight")
+
+            if len(galore_params) == 0:
+                raise ValueError(
+                    f"None of the target modules were found! ({args.optim_target_modules}). Please make sure to pass a valid `target_modules`."
+                )
+
+            non_galore_params = [p for n, p in model.named_parameters() if n not in galore_params_names]
+
+            galore_optim_kwargs = {
+                "rank": optim_args.pop("rank", 128),
+                "update_proj_gap": optim_args.pop("update_proj_gap", 200),
+                "scale": optim_args.pop("scale", 0.25),
+                "proj_type": optim_args.pop("proj_type", "std"),
+            }
+
+            # The default args are from the official repository: https://github.com/jiaweizzhao/GaLore
+            param_groups = [
+                {"params": non_galore_params},
+                {"params": galore_params, **galore_optim_kwargs},
+            ]
+
+            if is_layerwise:
+                # For layer-wise optimizers, the optimization step is done through post accumulation
+                # gradient hooks. The trick is to first attach these hooks to the model parameters then
+                # create a dummy optimizer that will perform no-ops in the Trainer.
+                # See the original implementation or the nice implementation from @hiyouga
+                # here: https://github.com/hiyouga/LLaMA-Factory/commit/8664262cde3919e10eaecbd66e8c5d356856362e#diff-ebe08ab14496dfb9e06075f0fdd36799ef6d1535cc4dd4715b74c4e3e06fe3ba
+                if args.gradient_accumulation_steps != 1:
+                    raise ValueError("Layerwise GaLoRE optimizer do not support gradient accumulation !")
+
+                optimizer_dict = {}
+                for param in non_galore_params:
+                    param_groups = [{"params": [param]}]
+                    optimizer_dict[param] = optimizer_cls(param_groups, **optimizer_kwargs)
+                for param in galore_params:
+                    param_groups = [{"params": [param], **galore_optim_kwargs}]
+                    optimizer_dict[param] = optimizer_cls(param_groups, **optimizer_kwargs)
+
+                def optimizer_hook(param):
+                    if param.grad is not None:
+                        optimizer_dict[param].step()
+                        optimizer_dict[param].zero_grad()
+
+                for param in model.parameters():
+                    param.register_post_accumulate_grad_hook(optimizer_hook)
+
+                optimizer_cls = LayerWiseDummyOptimizer
+                optimizer_kwargs.update({"optimizer_dict": optimizer_dict})
```

Comment:
All of this looks super hacky. At first glance, it looks okay, but I could imagine that some assumptions will fail in certain edge cases. Not sure what can really be done, except to try to cover as many use cases as possible (lr scheduler, checkpointing, gradient clipping, ...) in tests. Those tests would have to be a bit more involved than just testing that there is no error in training. AFAICT, the referenced repo does not perform these tests.

I guess this would be a lot of extra work, so feel free to ignore my comment and then we just cross our fingers that it works :)

Reply:
Yeah I agree ! For the other solution I guess we have to change the training logic as you need to somehow guard `optimizer.step()` and call it only if you don't have hooks attached :/ 
@hiyouga has perhaps more experience with this approach, as you say it's probably fine to keep it like this as layer-wise optimization is experimental and we can iterate in the future to see if it's stable with other features (thinking of DS, etc). I'll add a note on the docs about it

Reply:
I see thanks for explaining ! so I think we should just keep in mind this is temporary and we might adopt our approach according to how things will move in the future